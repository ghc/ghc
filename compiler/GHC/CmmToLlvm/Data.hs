
-- ----------------------------------------------------------------------------
-- | Handle conversion of CmmData to LLVM code.
--

module GHC.CmmToLlvm.Data (
        genLlvmData, genData
    ) where

import GHC.Prelude

import GHC.Llvm
import GHC.Llvm.Types (widenFp)
import GHC.CmmToLlvm.Base
import GHC.CmmToLlvm.Config

import GHC.Cmm.BlockId
import GHC.Cmm.CLabel
import GHC.Cmm.InitFini
import GHC.Cmm
import GHC.Platform

import GHC.Data.FastString
import GHC.Utils.Panic
import qualified Data.ByteString as BS

-- ----------------------------------------------------------------------------
-- * Constants
--

-- | The string appended to a variable name to create its structure type alias
structStr :: LMString
structStr = fsLit "_struct"

-- | The LLVM visibility of the label
linkage :: CLabel -> LlvmLinkageType
linkage lbl = if externallyVisibleCLabel lbl
              then ExternallyVisible else Internal

-- ----------------------------------------------------------------------------
-- * Top level
--

-- | Pass a CmmStatic section to an equivalent Llvm code.
genLlvmData :: (Section, RawCmmStatics) -> LlvmM LlvmData
-- See Note [emit-time elimination of static indirections] in "GHC.Cmm.CLabel".
genLlvmData (_, CmmStaticsRaw alias [CmmStaticLit (CmmLabel lbl), CmmStaticLit ind, _, _])
  | lbl == mkIndStaticInfoLabel
  , let labelInd (CmmLabelOff l _) = Just l
        labelInd (CmmLabel l) = Just l
        labelInd _ = Nothing
  , Just ind' <- labelInd ind
  , alias `mayRedirectTo` ind' = do
    label <- strCLabel_llvm alias
    label' <- strCLabel_llvm ind'
    let link     = linkage alias
        link'    = linkage ind'
        -- the LLVM type we give the alias is an empty struct type
        -- but it doesn't really matter, as the pointer is only
        -- used for (bit/int)casting.
        tyAlias  = LMAlias (label `appendFS` structStr, LMStructU [])

        aliasDef = LMGlobalVar label tyAlias link Nothing Nothing Alias
        -- we don't know the type of the indirectee here
        indType  = panic "will be filled by 'aliasify', later"
        orig     = LMStaticPointer $ LMGlobalVar label' indType link' Nothing Nothing Alias

    pure ([LMGlobal aliasDef $ Just orig], [tyAlias])

-- See Note [Initializers and finalizers in Cmm] in GHC.Cmm.InitFini.
genLlvmData (sect, statics)
  | Just (initOrFini, clbls) <- isInitOrFiniArray (CmmData sect statics)
  = let var = case initOrFini of
                IsInitArray -> fsLit "llvm.global_ctors"
                IsFiniArray -> fsLit "llvm.global_dtors"
    in genGlobalLabelArray var clbls

genLlvmData (sec, CmmStaticsRaw lbl xs) = do
    label <- strCLabel_llvm lbl
    static <- mapM genData xs
    lmsec <- llvmSection sec
    platform <- getPlatform
    let types   = map getStatType static

        strucTy = LMStruct types
        tyAlias = LMAlias (label `appendFS` structStr, strucTy)

        struct         = Just $ LMStaticStruc static tyAlias
        link           = linkage lbl
        align          = case sec of
                            Section CString _ -> if (platformArch platform == ArchS390X)
                                                    then Just 2 else Just 1
                            Section Data _    -> Just $ platformWordSizeInBytes platform
                            _                 -> Nothing
        const          = if sectionProtection sec == ReadOnlySection
                            then Constant else Global
        varDef         = LMGlobalVar label tyAlias link lmsec align const
        globDef        = LMGlobal varDef struct

    return ([globDef], [tyAlias])

-- | Produce an initializer or finalizer array declaration.
-- See Note [Initializers and finalizers in Cmm] in GHC.Cmm.InitFini for
-- details.
genGlobalLabelArray :: FastString -> [CLabel] -> LlvmM LlvmData
genGlobalLabelArray var_nm clbls = do
    lbls <- mapM strCLabel_llvm clbls
    decls <- mapM mkFunDecl lbls
    let entries = map toArrayEntry lbls
        static = LMStaticArray entries arr_ty
        arr = LMGlobal arr_var (Just static)
    return ([arr], decls)
  where
    mkFunDecl :: LMString -> LlvmM LlvmType
    mkFunDecl fn_lbl = do
        let fn_ty = mkFunTy fn_lbl
        funInsert fn_lbl fn_ty
        return (fn_ty)

    toArrayEntry :: LMString -> LlvmStatic
    toArrayEntry fn_lbl =
        let fn_var = LMGlobalVar fn_lbl (LMPointer $ mkFunTy fn_lbl) Internal Nothing Nothing Global
            fn = LMStaticPointer fn_var
            null = LMStaticLit (LMNullLit i8Ptr)
            prio = LMStaticLit $ LMIntLit 0xffff i32
        in LMStaticStrucU [prio, fn, null] entry_ty

    arr_var = LMGlobalVar var_nm arr_ty Appending Nothing Nothing Global
    mkFunTy lbl = LMFunction $ LlvmFunctionDecl lbl ExternallyVisible CC_Ccc LMVoid FixedArgs [] Nothing
    entry_ty = LMStructU [i32, LMPointer $ mkFunTy $ fsLit "placeholder", LMPointer i8]
    arr_ty = LMArray (length clbls) entry_ty

-- | Format the section type part of a Cmm Section
llvmSectionType :: Platform -> SectionType -> FastString
llvmSectionType p t = case t of
    Text                    -> fsLit ".text"
    ReadOnlyData            -> case platformOS p of
                                 OSMinGW32 -> fsLit ".rdata"
                                 _         -> fsLit ".rodata"
    RelocatableReadOnlyData -> case platformOS p of
                                 OSMinGW32 -> fsLit ".rdata$rel.ro"
                                 _         -> fsLit ".data.rel.ro"
    Data                    -> fsLit ".data"
    UninitialisedData       -> fsLit ".bss"
    CString                 -> case platformOS p of
                                 OSMinGW32 -> fsLit ".rdata$str"
                                 _         -> fsLit ".rodata.str"

    InitArray               -> panic "llvmSectionType: InitArray"
    FiniArray               -> panic "llvmSectionType: FiniArray"
    OtherSection _          -> panic "llvmSectionType: unknown section type"

-- | Format a Cmm Section into a LLVM section name
llvmSection :: Section -> LlvmM LMSection
llvmSection (Section t suffix) = do
  opts <- getConfig
  let splitSect = llvmCgSplitSection opts
      platform  = llvmCgPlatform     opts
  if not splitSect
  then return Nothing
  else do
    lmsuffix <- strCLabel_llvm suffix
    let result sep = Just (concatFS [llvmSectionType platform t
                                    , fsLit sep, lmsuffix])
    case platformOS platform of
      OSMinGW32 -> return (result "$")
      _         -> return (result ".")

-- ----------------------------------------------------------------------------
-- * Generate static data
--

-- | Handle static data
genData :: CmmStatic -> LlvmM LlvmStatic

genData (CmmFileEmbed {}) = panic "Unexpected CmmFileEmbed literal"
genData (CmmString str) = do
    let v  = map (\x -> LMStaticLit $ LMIntLit (fromIntegral x) i8)
                 (BS.unpack str)
        ve = v ++ [LMStaticLit $ LMIntLit 0 i8]
    return $ LMStaticArray ve (LMArray (length ve) i8)

genData (CmmUninitialised bytes)
    = return $ LMUninitType (LMArray bytes i8)

genData (CmmStaticLit lit)
    = genStaticLit lit

-- | Generate Llvm code for a static literal.
--
-- Will either generate the code or leave it unresolved if it is a 'CLabel'
-- which isn't yet known.
genStaticLit :: CmmLit -> LlvmM LlvmStatic
genStaticLit (CmmInt i w)
    = return $ LMStaticLit (LMIntLit i (LMInt $ widthInBits w))

genStaticLit (CmmFloat r W32)
    = return $ LMStaticLit (LMFloatLit (widenFp (fromRational r :: Float)) (widthToLlvmFloat W32))

genStaticLit (CmmFloat r W64)
    = return $ LMStaticLit (LMFloatLit (fromRational r :: Double) (widthToLlvmFloat W64))

genStaticLit (CmmFloat _r _w)
    = panic "genStaticLit (CmmLit:CmmFloat), unsupported float lit"

genStaticLit (CmmVec ls)
    = do sls <- mapM toLlvmLit ls
         return $ LMStaticLit (LMVectorLit sls)
  where
    toLlvmLit :: CmmLit -> LlvmM LlvmLit
    toLlvmLit lit = do
      slit <- genStaticLit lit
      case slit of
        LMStaticLit llvmLit -> return llvmLit
        _ -> panic "genStaticLit"

-- Leave unresolved, will fix later
genStaticLit cmm@(CmmLabel l) = do
    var <- getGlobalPtr =<< strCLabel_llvm l
    platform <- getPlatform
    let ptr = LMStaticPointer var
        lmty = cmmToLlvmType $ cmmLitType platform cmm
    return $ LMPtoI ptr lmty

genStaticLit (CmmLabelOff label off) = do
    platform <- getPlatform
    var <- genStaticLit (CmmLabel label)
    let offset = LMStaticLit $ LMIntLit (toInteger off) (llvmWord platform)
    return $ LMAdd var offset

genStaticLit (CmmLabelDiffOff l1 l2 off w) = do
    platform <- getPlatform
    var1 <- genStaticLit (CmmLabel l1)
    var2 <- genStaticLit (CmmLabel l2)
    let var
          | w == wordWidth platform = LMSub var1 var2
          | otherwise = LMTrunc (LMSub var1 var2) (widthToLlvmInt w)
        offset = LMStaticLit $ LMIntLit (toInteger off) (LMInt $ widthInBits w)
    return $ LMAdd var offset

genStaticLit (CmmBlock b) = genStaticLit $ CmmLabel $ infoTblLbl b

genStaticLit (CmmHighStackMark)
    = panic "genStaticLit: CmmHighStackMark unsupported!"
