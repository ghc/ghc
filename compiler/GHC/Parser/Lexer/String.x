{
{- |
This module defines lex states for strings.

This needs to be separate from the normal lexer because the normal lexer
automatically includes rules like skipping whitespace or lexing comments,
which we don't want in these contexts.
-}
module GHC.Parser.Lexer.String (
  AlexReturn (..),
  alexScan,
  string_multi_content,
) where

import GHC.Prelude

import GHC.Parser.Lexer.Interface
import GHC.Utils.Panic (panic)
}

-- -----------------------------------------------------------------------------
-- Alex "Character set macros"
-- Copied from GHC/Parser/Lexer.x

$unispace    = \x05 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$nl          = [\n\r\f]
$space       = [\ $unispace]
$whitechar   = [$nl \v $space]
$tab         = \t

$ascdigit  = 0-9
$unidigit  = \x03 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$decdigit  = $ascdigit -- exactly $ascdigit, no more no less.
$digit     = [$ascdigit $unidigit]

$special   = [\(\)\,\;\[\]\`\{\}]
$ascsymbol = [\!\#\$\%\&\*\+\.\/\<\=\>\?\@\\\^\|\-\~\:]
$unisymbol = \x04 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$symbol    = [$ascsymbol $unisymbol] # [$special \_\"\']

$unilarge  = \x01 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$asclarge  = [A-Z]
$large     = [$asclarge $unilarge]

$unismall  = \x02 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$ascsmall  = [a-z]
$small     = [$ascsmall $unismall \_]

$uniidchar = \x07 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$idchar    = [$small $large $digit $uniidchar \']

$unigraphic = \x06 -- Trick Alex into handling Unicode. See Note [Unicode in Alex].
$graphic   = [$small $large $symbol $digit $idchar $special $unigraphic \"\']
$charesc   = [a b f n r t v \\ \" \' \&]

$octit     = 0-7
$hexit     = [$decdigit A-F a-f]

-- -----------------------------------------------------------------------------
-- Alex "Regular expression macros"
-- Copied from GHC/Parser/Lexer.x

@numspc       = _*                   -- numeric spacer (#14473)
@decimal      = $decdigit(@numspc $decdigit)*
@octal        = $octit(@numspc $octit)*
@hexadecimal  = $hexit(@numspc $hexit)*
@gap = \\ $whitechar+ \\
@cntrl = $asclarge | \@ | \[ | \\ | \] | \^ | \_
@ascii = \^ @cntrl | "NUL" | "SOH" | "STX" | "ETX" | "EOT" | "ENQ" | "ACK"
       | "BEL" | "BS" | "HT" | "LF" | "VT" | "FF" | "CR" | "SO" | "SI" | "DLE"
       | "DC1" | "DC2" | "DC3" | "DC4" | "NAK" | "SYN" | "ETB" | "CAN"
       | "EM" | "SUB" | "ESC" | "FS" | "GS" | "RS" | "US" | "SP" | "DEL"
@escape     = \\ ( $charesc      | @ascii | @decimal | o @octal | x @hexadecimal )
@stringchar = ($graphic # [\\ \"]) | $space | @escape     | @gap

:-

-- Define an empty rule so it compiles; callers should always explicitly specify a startcode
<0> () ;

-- See Note [Lexing multiline strings]
<string_multi_content> {
  -- Parse as much of the multiline string as possible, except for quotes
  @stringchar* ($nl ([\  $tab] | @gap)* @stringchar*)* { string_multi_content_action }
  -- Allow bare quotes if it's not a triple quote
  (\" | \"\") / ([\n .] # \") { string_multi_content_action }
}

-- -----------------------------------------------------------------------------
-- Haskell actions
{
-- | Dummy action that should never be called. Should only be used in lex states
-- that are manually lexed in tok_string_multi.
string_multi_content_action :: a
string_multi_content_action = panic "string_multi_content_action unexpectedly invoked"
}
