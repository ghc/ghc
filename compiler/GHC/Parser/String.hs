{-# LANGUAGE LambdaCase #-}

module GHC.Parser.String (
  StringLexError (..),
  ContainsSmartQuote (..),
  LexStringType (..),
  lexString,

  -- * Unicode smart quote helpers
  isDoubleSmartQuote,
  isSingleSmartQuote,

  -- * Other helpers
  isAnyChar,
  resolveEscapeCharacter,
) where

import GHC.Prelude

import Control.Arrow ((>>>))
import Control.Monad (guard, unless, when)
import Data.Char (chr, isPrint, ord)
import Data.List (unfoldr)
import Data.Maybe (listToMaybe, mapMaybe)
import GHC.Parser.CharClass (
  hexDigit,
  is_any,
  is_decdigit,
  is_hexdigit,
  is_octdigit,
  is_space,
  octDecDigit,
 )
import GHC.Utils.Panic (panic)

data LexStringType = StringTypeSingle | StringTypeMulti

-- | State to accumulate while iterating through string literal.
--
-- Fields are strict here to avoid space leak when iterating through large string literal
-- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12089#note_576175
data LexStringState loc = LexStringState
  { stringAcc :: !String
    -- ^ The string seen so far, reversed
  , multilineCommonWsPrefix :: !Int
    -- ^ The common prefix for multiline strings. See Note [Multiline string literals]
  , initialLoc :: !loc
    -- ^ The location of the beginning of the string literal
  }

-- | Get the character at the given location, with the location
-- of the next character. Returns Nothing if at the end of the
-- input.
type GetChar loc = loc -> Maybe (Char, loc)

lexString :: LexStringType -> GetChar loc -> loc -> Either (StringLexError loc) (String, loc)
lexString strType getChar initialLoc = go initialState initialLoc
  where
    initialState =
      LexStringState
        { stringAcc = ""
        , multilineCommonWsPrefix =
            case strType of
              StringTypeMulti -> maxBound
              _ -> 0
        , initialLoc = initialLoc
        }

    -- 's' is strict here to avoid space leak when iterating through large string literal
    -- https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12089#note_576175
    go !s loc0 =
      case getChar loc0 of
        -- found closing delimiter
        Just ('"', _) | Just loc1 <- checkDelimiter strType getChar loc0 -> do
          let postprocess =
                case strType of
                  StringTypeSingle -> id
                  StringTypeMulti -> postprocessMultiline (multilineCommonWsPrefix s)
          Right (postprocess . reverse $ stringAcc s, loc1)

        -- found backslash
        Just (c0@'\\', loc1) -> do
          case getChar loc1 of
            -- found '\&' character, which should be elided
            Just ('&', loc2) -> go s loc2
            -- found start of a string gap
            Just (c1, loc2) | is_space c1 -> collapseStringGap getChar s loc2 >>= go s
            -- some other escape character
            Just (c1, loc2) ->
              case strType of
                StringTypeSingle -> do
                  (c', loc') <- resolveEscapeCharacter getChar loc1
                  go (addChar c' s) loc'
                StringTypeMulti -> do
                  -- keep escape characters unresolved until after post-processing,
                  -- to distinguish between a user-newline and the user writing "\n".
                  -- but still process the characters here, to find any errors
                  _ <- resolveEscapeCharacter getChar loc1
                  go (addChar c1 . addChar c0 $ s) loc2
            -- backslash at end of input
            Nothing -> Left $ BadCharInitialLex loc1 (hasSQuote getChar s)

        -- found newline character in multiline string
        Just (c0@'\n', loc1) | StringTypeMulti <- strType ->
          uncurry go $ parseLeadingWS getChar (addChar c0 s) loc1

        -- found some other character
        Just (c0, loc1) | isAnyChar c0 -> go (addChar c0 s) loc1

        -- found some unknown character
        Just (_, _) -> Left $ BadCharInitialLex loc0 (hasSQuote getChar s)

        -- reached EOF before finding end of string
        Nothing -> Left $ BadCharInitialLex loc0 (hasSQuote getChar s)
{-# INLINE lexString #-}

checkDelimiter :: LexStringType -> GetChar loc -> loc -> Maybe loc
checkDelimiter strType getChar loc0 =
  case strType of
    StringTypeSingle -> do
      ('"', loc1) <- getChar loc0
      Just loc1
    StringTypeMulti -> do
      ('"', loc1) <- getChar loc0
      ('"', loc2) <- getChar loc1
      ('"', loc3) <- getChar loc2
      Just loc3
{-# INLINE checkDelimiter #-}

-- | A helper for adding the given character to the lexed string.
addChar :: Char -> LexStringState loc -> LexStringState loc
addChar c s = s{stringAcc = c : stringAcc s}
{-# INLINE addChar #-}

-- | Return whether the string we've parsed so far contains any smart quotes.
hasSQuote :: GetChar loc -> LexStringState loc -> ContainsSmartQuote loc
hasSQuote getChar s
  | any isDoubleSmartQuote (stringAcc s)
  , (c, loc) : _ <- filter (isDoubleSmartQuote . fst) allChars =
      SmartQuote c loc
  | otherwise =
      NoSmartQuote
  where
    allChars = unfoldr getCharWithLoc (initialLoc s)
    getCharWithLoc loc =
      case getChar loc of
        Just (c, loc') -> Just ((c, loc), loc')
        Nothing -> Nothing
{-# INLINE hasSQuote #-}

-- | After parsing a backslash and a space character, consume the rest of
-- the string gap and return the next location.
collapseStringGap :: GetChar loc -> LexStringState loc -> loc -> Either (StringLexError loc) loc
collapseStringGap getChar s = go
  where
    go loc0 =
      case getChar loc0 of
        Just ('\\', loc1) -> pure loc1
        Just (c0, loc1) | is_space c0 -> go loc1
        Just _ -> Left $ BadCharInitialLex loc0 (hasSQuote getChar s)
        Nothing -> Left $ UnexpectedEOF loc0 (hasSQuote getChar s)
{-# INLINE collapseStringGap #-}

-- | See Note [Multiline string literals]
parseLeadingWS :: GetChar loc -> LexStringState loc -> loc -> (LexStringState loc, loc)
parseLeadingWS getChar = go 0
  where
    go !col s loc =
      case getChar loc of
        Just (c@' ', loc') -> go (col + 1) (addChar c s) loc'
        -- expand tabs
        Just ('\t', loc') ->
          let fill = 8 - (col `mod` 8)
              s' = applyN fill (addChar ' ') s
           in go (col + fill) s' loc'
        -- if we see a newline or string delimiter, then this line only contained whitespace, so
        -- don't include it in the common whitespace prefix
        Just ('\n', _) -> (s, loc)
        Just ('"', _) | Just _ <- checkDelimiter StringTypeMulti getChar loc -> (s, loc)
        -- found some other character, so we're done parsing leading whitespace
        _ ->
          let s' = s{multilineCommonWsPrefix = min col (multilineCommonWsPrefix s)}
           in (s', loc)

    applyN :: Int -> (a -> a) -> a -> a
    applyN n f x0 = iterate f x0 !! n
{-# INLINE parseLeadingWS #-}

data StringLexError loc
  = UnexpectedEOF !loc !(ContainsSmartQuote loc)
    -- ^ Unexpectedly hit EOF when lexing string
  | BadCharInitialLex !loc !(ContainsSmartQuote loc)
    -- ^ Found invalid character when initially lexing string
  | EscapeBadChar !loc
    -- ^ Found invalid character when parsing an escaped character
  | EscapeUnexpectedEOF !loc
    -- ^ Unexpectedly hit EOF when parsing an escaped character
  | EscapeNumRangeError !loc
    -- ^ Escaped number exceeds range
  | EscapeSmartQuoteError !Char !loc
    -- ^ Found escaped smart unicode chars as `\’` or `\”`
  deriving (Show)

-- | When initially lexing the string, we want to track if we've
-- seen a smart quote, to show a helpful "you might be accidentally
-- using a smart quote" error.
data ContainsSmartQuote loc
  = NoSmartQuote
  | SmartQuote !Char !loc
  deriving (Show)

-- -----------------------------------------------------------------------------
-- Escape characters

-- | After finding a backslash, parse the rest of the escape character, starting
-- at the given location.
resolveEscapeCharacter :: GetChar loc -> loc -> Either (StringLexError loc) (Char, loc)
resolveEscapeCharacter getChar loc0 = do
  (c0, loc1) <- expectChar loc0
  case c0 of
    'a'  -> pure ('\a', loc1)
    'b'  -> pure ('\b', loc1)
    'f'  -> pure ('\f', loc1)
    'n'  -> pure ('\n', loc1)
    'r'  -> pure ('\r', loc1)
    't'  -> pure ('\t', loc1)
    'v'  -> pure ('\v', loc1)
    '\\' -> pure ('\\', loc1)
    '"'  -> pure ('\"', loc1)
    '\'' -> pure ('\'', loc1)
    -- escape codes
    'x' -> expectNum is_hexdigit 16 hexDigit loc1
    'o' -> expectNum is_octdigit 8 octDecDigit loc1
    _ | is_decdigit c0 -> expectNum is_decdigit 10 octDecDigit loc0
    -- control characters (e.g. '\^M')
    '^' -> do
      (c1, loc2) <- expectChar loc1
      unless ('@' <= c1 && c1 <= '_') $ Left $ EscapeBadChar loc1
      pure (chr $ ord c1 - ord '@', loc2)
    -- long form escapes (e.g. '\NUL')
    _ | Just (c1, loc2) <- parseLongEscape getChar c0 loc1 -> pure (c1, loc2)
    -- check unicode smart quotes (#21843)
    _ | isDoubleSmartQuote c0 -> Left $ EscapeSmartQuoteError c0 loc0
    _ | isSingleSmartQuote c0 -> Left $ EscapeSmartQuoteError c0 loc0
    -- unknown escape
    _ -> Left $ EscapeBadChar loc0
  where
    expectChar loc =
      case getChar loc of
        Just x -> pure x
        Nothing -> Left $ EscapeUnexpectedEOF loc

    expectNum isDigit base toDigit loc1 = do
      (c1, loc2) <- expectChar loc1
      unless (isDigit c1) $ Left $ EscapeBadChar loc1
      let parseNum x loc =
            case getChar loc of
              Just (c, loc') | isDigit c -> do
                let x' = x * base + toDigit c
                when (x' > 0x10ffff) $ Left $ EscapeNumRangeError loc
                parseNum x' loc'
              _ ->
                pure (chr x, loc)
      parseNum (toDigit c1) loc2
{-# INLINE resolveEscapeCharacter #-}

parseLongEscape :: GetChar loc -> Char -> loc -> Maybe (Char, loc)
parseLongEscape getChar c0 loc1 = listToMaybe $ mapMaybe tryParse longEscapeCodes
  where
    tryParse (prefix, c) = do
      p0 : p <- pure prefix
      guard (p0 == c0)          -- see if the first character matches
      loc <- parsePrefix loc1 p -- see if the rest of the prefix matches
      pure (c, loc)

    parsePrefix loc = \case
      [] -> pure loc
      p : ps -> do
        (c, loc') <- getChar loc
        guard (p == c)
        parsePrefix loc' ps

    longEscapeCodes =
      [ ("NUL", '\NUL')
      , ("SOH", '\SOH')
      , ("STX", '\STX')
      , ("ETX", '\ETX')
      , ("EOT", '\EOT')
      , ("ENQ", '\ENQ')
      , ("ACK", '\ACK')
      , ("BEL", '\BEL')
      , ("BS", '\BS')
      , ("HT", '\HT')
      , ("LF", '\LF')
      , ("VT", '\VT')
      , ("FF", '\FF')
      , ("CR", '\CR')
      , ("SO", '\SO')
      , ("SI", '\SI')
      , ("DLE", '\DLE')
      , ("DC1", '\DC1')
      , ("DC2", '\DC2')
      , ("DC3", '\DC3')
      , ("DC4", '\DC4')
      , ("NAK", '\NAK')
      , ("SYN", '\SYN')
      , ("ETB", '\ETB')
      , ("CAN", '\CAN')
      , ("EM", '\EM')
      , ("SUB", '\SUB')
      , ("ESC", '\ESC')
      , ("FS", '\FS')
      , ("GS", '\GS')
      , ("RS", '\RS')
      , ("US", '\US')
      , ("SP", '\SP')
      , ("DEL", '\DEL')
      ]
{-# INLINE parseLongEscape #-}

-- -----------------------------------------------------------------------------
-- Unicode Smart Quote detection (#21843)

isDoubleSmartQuote :: Char -> Bool
isDoubleSmartQuote = \case
  '“' -> True
  '”' -> True
  _ -> False

isSingleSmartQuote :: Char -> Bool
isSingleSmartQuote = \case
  '‘' -> True
  '’' -> True
  _ -> False

{-
Note [Multiline string literals]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Multiline string literals were added following the acceptance of the
proposal: https://github.com/ghc-proposals/ghc-proposals/pull/569

Multiline string literals are syntax sugar for normal string literals,
with an extra post processing step. This all happens in the Lexer; that
is, HsMultilineString will contain the post-processed string. This matches
the same behavior as HsString, which contains the normalized string
(see Note [Literal source text]).

The canonical steps for post processing a multiline string are:
1. Collapse string gaps
2. Split the string by newlines
3. Convert leading tabs into spaces
    * In each line, any tabs preceding non-whitespace characters are replaced with spaces up to the next tab stop
4. Remove common whitespace prefix in every line (see below)
5. If a line contains only whitespace, remove all of the whitespace
6. Join the string back with `\n` delimiters
7. If the first character of the string is a newline, remove it
8. Interpret escaped characters

However, for performance reasons, we do as much of this in one pass as possible:
1. As we lex the string, do the following steps as they appear:
    a. Collapse string gaps
    b. Keep track of the common whitespace prefix so far
    c. Validate escaped characters
2. At the very end, post process the lexed string:
    a. Remove the common whitespace prefix from every line
    b. Remove all whitespace from all-whitespace lines
    c. Remove initial newline character
    d. Resolve escaped characters

The common whitespace prefix can be informally defined as "The longest
prefix of whitespace shared by all lines in the string, excluding the
first line and any whitespace-only lines".

It's more precisely defined with the following algorithm:

1. Take a list representing the lines in the string
2. Ignore the following elements in the list:
    * The first line (we want to ignore everything before the first newline)
    * Empty lines
    * Lines with only whitespace characters
3. Calculate the longest prefix of whitespace shared by all lines in the remaining list
-}

-- | See Note [Multiline string literals]
postprocessMultiline :: Int -> String -> String
postprocessMultiline commonWSPrefix =
      rmCommonWhitespacePrefix
  >>> collapseOnlyWsLines
  >>> rmFirstNewline
  >>> rmLastNewline
  >>> resolveEscapeChars
  where
    rmCommonWhitespacePrefix =
      let go = \case
            '\n' : s -> '\n' : go (dropLine commonWSPrefix s)
            c : s -> c : go s
            [] -> []
          -- drop x characters from the string, or up to a newline, whichever
          -- comes first
          dropLine !x = \case
            s | x <= 0 -> s
            s@('\n' : _) -> s
            _ : s -> dropLine (x - 1) s
            [] -> []
       in go

    collapseOnlyWsLines =
      let go = \case
            '\n' : s | Just s' <- checkAllWs s -> '\n' : go s'
            c : s -> c : go s
            [] -> []
          checkAllWs = \case
            -- got all the way to a newline or the end of the string, return
            s@('\n' : _) -> Just s
            s@[] -> Just s
            -- found whitespace, continue
            c : s | is_space c -> checkAllWs s
            -- anything else, stop
            _ -> Nothing
       in go

    rmFirstNewline = \case
      '\n' : s -> s
      s -> s

    rmLastNewline =
      let go = \case
            [] -> []
            ['\n'] -> []
            c : cs -> c : go cs
       in go

    -- resolve escape characters, deferred from lexString. guaranteed
    -- to not throw any errors, since we already checked them in lexString
    resolveEscapeChars = \case
      [] -> []
      '\\' : s ->
        -- concretizing 'loc' to String:
        --   resolveEscapeCharacter :: (String -> Maybe (Char, String)) -> String -> Either _ (Char, String)
        case resolveEscapeCharacter uncons s of
          Left e -> panic $ "resolving escape characters in multiline string unexpectedly found errors: " ++ show e
          Right (c, s') -> c : resolveEscapeChars s'
      c : s -> c : resolveEscapeChars s

    uncons = \case
      c : cs -> Just (c, cs)
      [] -> Nothing

-- -----------------------------------------------------------------------------
-- Helpers

isAnyChar :: Char -> Bool
isAnyChar c
  | c > '\x7f' = isPrint c
  | otherwise  = is_any c
