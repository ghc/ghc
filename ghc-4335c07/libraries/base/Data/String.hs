{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE TypeFamilies #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.String
-- Copyright   :  (c) The University of Glasgow 2007
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  experimental
-- Portability :  portable
--
-- The @String@ type and associated operations.
--
-----------------------------------------------------------------------------

module Data.String (
   String
 , IsString(..)

 -- * Functions on strings
 , lines
 , words
 , unlines
 , unwords
 ) where

import GHC.Base
import Data.Functor.Const (Const (Const))
import Data.Functor.Identity (Identity (Identity))
import Data.List (lines, words, unlines, unwords)

-- | Class for string-like datastructures; used by the overloaded string
--   extension (-XOverloadedStrings in GHC).
class IsString a where
    fromString :: String -> a

{- Note [IsString String]
~~~~~~~~~~~~~~~~~~~~~~~~~
Previously, the IsString instance that covered String was a flexible
instance for [Char]. This is in some sense the most accurate choice,
but there are cases where it can lead to an ambiguity, for instance:

  show $ "foo" ++ "bar"

The use of (++) ensures that "foo" and "bar" must have type [t] for
some t, but a flexible instance for [Char] will _only_ match if
something further determines t to be Char, and nothing in the above
example actually does.

So, the above example generates an error about the ambiguity of t,
and what's worse, the above behavior can be generated by simply
typing:

   "foo" ++ "bar"

into GHCi with the OverloadedStrings extension enabled.

The new instance fixes this by defining an instance that matches all
[a], and forces a to be Char. This instance, of course, overlaps
with things that the [Char] flexible instance doesn't, but this was
judged to be an acceptable cost, for the gain of providing a less
confusing experience for people experimenting with overloaded strings.

It may be possible to fix this via (extended) defaulting. Currently,
the rules are not able to default t to Char in the above example. If
a more flexible system that enabled this defaulting were put in place,
then it would probably make sense to revert to the flexible [Char]
instance, since extended defaulting is enabled in GHCi. However, it
is not clear at the time of this note exactly what such a system
would be, and it certainly hasn't been implemented.

A test case (should_run/overloadedstringsrun01.hs) has been added to
ensure the good behavior of the above example remains in the future.
-}

-- | @(a ~ Char)@ context was introduced in @4.9.0.0@
--
-- @since 2.01
instance (a ~ Char) => IsString [a] where
         -- See Note [IsString String]
    fromString xs = xs

-- | @since 4.9.0.0
deriving instance IsString a => IsString (Const a b)
deriving instance IsString a => IsString (Identity a)
