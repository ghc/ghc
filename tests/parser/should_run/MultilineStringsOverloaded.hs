{-# LANGUAGE MultilineStrings #-}
{-# LANGUAGE OverloadedStrings #-}

import Data.String (IsString (..))
import Data.Text (Text)

newtype Lines s = Lines [s]
  deriving (Show)

instance IsString s => IsString (Lines s) where
  fromString = Lines . map fromString . lines

lines0 :: Lines Text
lines0 =
  """
  this is
  a test
  with multiple lines
  """

main :: IO ()
main = print lines0
