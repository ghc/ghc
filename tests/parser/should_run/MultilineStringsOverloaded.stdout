Lines ["this is","a test","with multiple lines"]
